rmw_zenoh_cpp/src/detail/qos.hpp

```diff
@@ -0,0 +1,56 @@
+// Copyright 2024 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DETAIL__QOS_HPP_
+#define DETAIL__QOS_HPP_
+
+#include <rcutils/logging.h>
+#include <rmw/topic_endpoint_info_array.h>
+#include <rmw/types.h>
+
+#include <functional>
+
+namespace rmw_zenoh_cpp
+{
+//==============================================================================
+/// Signature matching rmw_get_publishers_info_by_topic and rmw_get_subscriptions_info_by_topic
+using GetEndpointInfoByTopicFunction = std::function<rmw_ret_t(
+      const rmw_node_t *,
+      rcutils_allocator_t *,
+      const char *,
+      bool,
+      rmw_topic_endpoint_info_array_t *)>;
+
+//==============================================================================
+class QoS
```

Comment:
This class feels awkward to me.

I wonder whether we need to have the class at all.  It seems to me that in `qos.cpp`, we could instead have:

```
static rmw_qos_profile_t default_qos =
{
  RMW_QOS_POLICY_HISTORY_SYSTEM_DEFAULT,
  42,
  RMW_QOS_POLICY_RELIABILITY_RELIABLE,
  RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL,
  RMW_DURATION_INFINITE,
  RMW_DURATION_INFINITE,
  RMW_QOS_POLICY_LIVELINESS_AUTOMATIC,
  RMW_DURATION_INFINITE,
  false,
};
```

And then have `best_available_qos` be a free function, something like:

```
rmw_ret_t best_available_qos(rmw_qos_profile_t * qos_profile)
{
  // pull data from default_qos as necessary
}
```

That said, I don't know exactly what your plans are for upcoming PRs.

Reply:
I was planning to update [qos_to_keyexpr](https://github.com/ros2/rmw_zenoh/blob/9ede38d1d84e9b7f08971d56d1843624b5052240/rmw_zenoh_cpp/src/detail/liveliness_utils.cpp#L174) to call `QoS()::get().default_qos()` and compare each field in the default with that of the input qos. Hence a class. I could also just define the `static rmw_qos_profile_t default_qos` as you mentioned here in the header file. Let me know if you have a preference.

Reply:
I see.  In that case you would have to export `rmw_qos_profile_t default_qos` anyway, since to re-implement `qos_to_keyexpr` you need access to each individual field.

Overall, even though I'm not a huge fan of this class, it isn't all that important in the end.  Thus I'm going to approve this anyway so we can get the improvements here.  We can always consider re-implementing this in the future if we think of a better way to do this.